File: frontend/src/db/licenses.ts
import { desc, eq } from "drizzle-orm";
import { getDb } from "@/db/index";
import { licenses, type LicenseRow } from "@/db/schema";
import type { CreateLicenseBody, License, MessageResponse } from "@/lib/types";

function toLicense(row: LicenseRow): License {
  return {
    license_key: row.licenseKey,
    product: row.product,
    status: row.status,
    machine_id: row.machineId,
    activated_at: row.activatedAt,
    created_at: row.createdAt,
    notes: row.notes,
  };
}

export async function listLicenses(): Promise<License[]> {
  const db = getDb();
  const rows = await db
    .select()
    .from(licenses)
    .orderBy(desc(licenses.createdAt));
  return rows.map(toLicense);
}

export async function createLicense(
  body: CreateLicenseBody
): Promise<MessageResponse> {
  const db = getDb();
  const licenseKey = body.license_key.trim();
  const product = body.product?.trim() || "default";

  const existing = await db
    .select({ id: licenses.id })
    .from(licenses)
    .where(eq(licenses.licenseKey, licenseKey))
    .limit(1);

  if (existing.length > 0) {
    return { success: false, message: "License key already exists." };
  }

  await db.insert(licenses).values({
    licenseKey,
    product,
    status: "inactive",
    notes: body.notes?.trim() || null,
  });

  return {
    success: true,
    message: "License key created.",
    data: { license_key: licenseKey },
  };
}

export async function revokeLicense(
  licenseKey: string
): Promise<MessageResponse> {
  const db = getDb();
  const key = licenseKey.trim();

  const existing = await db
    .select({ id: licenses.id })
    .from(licenses)
    .where(eq(licenses.licenseKey, key))
    .limit(1);

  if (existing.length === 0) {
    return { success: false, message: "License key not found." };
  }

  await db
    .update(licenses)
    .set({
      status: "inactive",
      machineId: null,
      activatedAt: null,
    })
    .where(eq(licenses.licenseKey, key));

  return {
    success: true,
    message: "License key revoked and reset to inactive.",
  };
}

File: backend/app/routers/licenses.py
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, Request
from sqlalchemy import select
from sqlalchemy.orm import Session
from slowapi import Limiter
from slowapi.util import get_remote_address

from app.database import get_db
from app.models.license import License
from app.models.schemas import (
    ActivateRequest,
    CreateLicenseRequest,
    LicenseResponse,
    MessageResponse,
    VerifyRequest,
)
from app.security import require_admin_key, require_api_key

router = APIRouter()
limiter = Limiter(key_func=get_remote_address)


def _to_response(license_row: License) -> LicenseResponse:
    return LicenseResponse(
        license_key=license_row.license_key,
        product=license_row.product,
        status=license_row.status,
        machine_id=license_row.machine_id,
        activated_at=(
            license_row.activated_at.isoformat() if license_row.activated_at else None
        ),
        created_at=license_row.created_at.isoformat(),
        notes=license_row.notes,
    )


def _utc_now() -> datetime:
    return datetime.now(timezone.utc)


# ── Client endpoints (used by your installer app) ───────────────

@router.post("/verify-key", response_model=MessageResponse)
@limiter.limit("10/minute")
def verify_key(
    request: Request,
    body: VerifyRequest,
    db: Session = Depends(get_db),
    _key=Depends(require_api_key),
):
    row = db.scalar(
        select(License).where(
            License.license_key == body.license_key,
            License.product == body.product,
        )
    )

    if not row:
        return MessageResponse(success=False, message="License key not found.")

    if row.status == "active":
        return MessageResponse(
            success=False,
            message="License key is already activated on another machine.",
            data={"status": row.status},
        )

    return MessageResponse(
        success=True,
        message="License key is valid and available.",
        data={"status": row.status},
    )


@router.post("/activate-key", response_model=MessageResponse)
@limiter.limit("5/minute")
def activate_key(
    request: Request,
    body: ActivateRequest,
    db: Session = Depends(get_db),
    _key=Depends(require_api_key),
):
    row = db.scalar(
        select(License).where(
            License.license_key == body.license_key,
            License.product == body.product,
        )
    )

    if not row:
        return MessageResponse(success=False, message="License key not found.")

    if row.status == "active":
        if row.machine_id == body.machine_id:
            return MessageResponse(
                success=True,
                message="License already active on this machine.",
                data={"status": "active"},
            )
        return MessageResponse(
            success=False,
            message="License key is already activated on a different machine.",
        )

    now = _utc_now()
    row.status = "active"
    row.machine_id = body.machine_id
    row.activated_at = now
    db.commit()

    return MessageResponse(
        success=True,
        message="License activated successfully.",
        data={"status": "active", "activated_at": now.isoformat()},
    )


# ── Admin endpoints (used by your key generator app) ───────────

@router.post("/admin/create-key", response_model=MessageResponse)
def create_key(
    body: CreateLicenseRequest,
    db: Session = Depends(get_db),
    _key=Depends(require_admin_key),
):
    existing = db.scalar(
        select(License.id).where(License.license_key == body.license_key)
    )

    if existing:
        return MessageResponse(success=False, message="License key already exists.")

    license_row = License(
        license_key=body.license_key,
        product=body.product or "default",
        status="inactive",
        notes=body.notes,
    )
    db.add(license_row)
    db.commit()

    return MessageResponse(
        success=True,
        message="License key created.",
        data={"license_key": body.license_key},
    )


@router.get("/admin/list-keys", response_model=list[LicenseResponse])
def list_keys(
    db: Session = Depends(get_db),
    _key=Depends(require_admin_key),
):
    rows = db.scalars(select(License).order_by(License.created_at.desc())).all()
    return [_to_response(row) for row in rows]


@router.delete("/admin/revoke-key/{license_key}", response_model=MessageResponse)
def revoke_key(
    license_key: str,
    db: Session = Depends(get_db),
    _key=Depends(require_admin_key),
):
    row = db.scalar(select(License).where(License.license_key == license_key))

    if not row:
        return MessageResponse(success=False, message="License key not found.")

    row.status = "inactive"
    row.machine_id = None
    row.activated_at = None
    db.commit()

    return MessageResponse(
        success=True,
        message="License key revoked and reset to inactive.",
    )
